home/migrations/0006_auto_20200623_2207.py

```diff
@@ -0,0 +1,254 @@
+# Generated by Django 3.0.5 on 2020-06-24 05:07
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django.utils.timezone
+import uuid
+
+
+def migrate_appusers(apps, schema_editor):
+    with schema_editor.connection.cursor() as cursor:
+        # create unique accounts for each email
+        cursor.execute("""
+            INSERT INTO home_account (email, name, zip, age, created, updated)
+            SELECT au.email, au.name, au.zip, au.age, au.created, NOW()
+            FROM home_appuser au
+            WHERE au.created=(
+                SELECT MIN(au2.created)
+                FROM home_appuser au2
+                WHERE LOWER(au2.email)=LOWER(au.email)
+            )
+        """)
+        # copy all of the device ids into the new devices table
+        cursor.execute("""
+            INSERT INTO home_device (device_id, account_id, created)
+            SELECT au.account_id, ac.id, au.created
+            FROM home_appuser au
+            JOIN home_account ac
+            ON LOWER(au.email)=LOWER(ac.email)
+        """)
+        # create some dummy rows to set initial default values in migration
+        cursor.execute("""
+            INSERT INTO home_account (id, email, name, zip, age, created, updated)
+            VALUES (0, 'dummy@email.com', 'Dummy Account', '94103', 0, NOW(), NOW())
+        """)
+        cursor.execute("""
+            INSERT INTO home_device (device_id, account_id, created)
+            VALUES ('0', 0, NOW())
+        """)
+
+
+def migrate_walks(apps, schema_editor):
+    with schema_editor.connection.cursor() as cursor:
```

Comment:
This function migrates all the walk data from referencing AppUser to Account/Device.

Actually, looking at this now- does each row need to reference BOTH Account and Device? Because referencing Device only would be enough to still be able to uniquely identify the Account (through a join on Device). Are you de-normalizing to make the query more efficient?

Reply:
Each walk only requires a device. On save, it automatically adds a direct reference to the account as well. I put this in explicitly so I can add the unique constraint on the account-date field and it was also simpler to reference the account in all the admin stuff. Granted, I could also simply write a function do to this as well but this seemed simplest (It looked like Django fields didn't allow reference resolution when configuring admin). This is mostly cosmetic in that it's only for unique constraint & admin. For any other functionality, its fairly easy to chain through the device to get to the account.